### PR TITLE
Add configurable runtime plugin to send_v2 so that it is usable

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
- `send_v2()` doesn't work for a very small number of reasons—allow it one last final runtime_plugin so that we can use it to make an e2e request

## Description
- update a few decorators to be deterministic
## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/rust-runtime/aws-runtime/src/invocation_id.rs [hunk 1]
```diff
@@ -45,11 +42,31 @@ impl Interceptor<HttpRequest, HttpResponse> for InvocationIdInterceptor {
...
 }
 
+#[derive(Debug)]
+pub struct InvocationId(HeaderValue);
+impl InvocationId {
+    pub fn for_tests() -> Self {
+        InvocationId(HeaderValue::from_static(
+            "000000000000000000000000000000000000",
+        ))
+    }
```
Comment (Velfi) [inline aws/rust-runtime/aws-runtime/src/invocation_id.rs:58]: Thanks for adding this.

File: aws/rust-runtime/aws-runtime/src/auth.rs [hunk 2]
```diff
@@ -196,6 +196,7 @@ pub mod sigv4 {
             &self,
             request: &mut HttpRequest,
             identity: &Identity,
+            // TODO: should this be the config bag?
```
Comment (jdisanti) [inline aws/rust-runtime/aws-runtime/src/auth.rs:199]:
```suggestion
            // TODO(enableNewSmithyRuntime): should this be the config bag?
```

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4AuthDecorator.kt [hunk 3]
```diff
@@ -136,6 +136,7 @@ private class AuthOperationRuntimePluginCustomization(private val codegenContext
                         signing_options.normalize_uri_path = $normalizeUrlPath;
                         signing_options.signing_optional = $signingOptional;
                         signing_options.payload_override = #{payload_override};
+                        println!("{:?}", cfg.request_time().unwrap());
```
Comment (jdisanti) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4AuthDecorator.kt:139]:
```suggestion
```

File: aws/sra-test/integration-tests/aws-sdk-s3/tests/sra_test.rs [hunk 4]
```diff
@@ -3,280 +3,98 @@
...
+            .bucket("test-bucket");
+
+        cfg.put(params_builder);
+        cfg.set_request_time(RequestTime::new(self.timestamp.clone()));
+        cfg.put(AwsUserAgent::for_tests());
+        cfg.put(InvocationId::for_tests());
+        Ok(())
     }
-
-    let conn = TestConnection::new(vec![(
```
Comment (jdisanti) [inline aws/sra-test/integration-tests/aws-sdk-s3/tests/sra_test.rs:219]: I would prefer to keep this test using a pre-recorded request/response.

**Review by jdisanti — COMMENTED** [review]
This is awesome!

File: aws/sra-test/integration-tests/aws-sdk-s3/tests/sra_test.rs (hunk 4 above)
Comment (rcoh) [reply aws/sra-test/integration-tests/aws-sdk-s3/tests/sra_test.rs:219]: it is still prerecorded but I changed it to use DVR, will make it easier to add more

File: aws/sra-test/integration-tests/aws-sdk-s3/test-data/list-objects-v2.json [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      "connection_id": 0,
+      "action": {
+        "Eof": {
+          "ok": true,
+          "direction": "Response"
+        }
+      }
+    }
+  ],
+  "docs": "todo docs",
```
Comment (Velfi) [inline aws/sra-test/integration-tests/aws-sdk-s3/test-data/list-objects-v2.json:103]: This field is meant to document what this tests, right? Can we set it to something like this for now?
```suggestion
  "docs": "Test sending an S3 ListObjectsV2 operation with a successful response.",
```

File: aws/sra-test/integration-tests/aws-sdk-s3/tests/sra_test.rs [hunk 6]
```diff
@@ -3,282 +3,75 @@
...
 
-    let _ = dbg!(
+    let resp = dbg!(
         client
             .list_objects_v2()
             .config_override(aws_sdk_s3::Config::builder().force_path_style(false))
             .bucket("test-bucket")
             .prefix("prefix~")
-            .send_v2()
+            .send_v2(Some(fixup))
```
Comment (Velfi) [inline aws/sra-test/integration-tests/aws-sdk-s3/tests/sra_test.rs:48]: I could see us adding a version of the send method like this that's only compiled for tests. I don't like the idea of users having to specify `None` for most requests though.

```rust
#[cfg(test)]
send_with_plugin(plugin: RuntimePlugin) -> Output { ... }
```

File: rust-runtime/aws-smithy-client/src/dvr/record.rs [hunk 7]
```diff
@@ -66,6 +70,14 @@ impl<S> RecordingConnection<S> {
         }
     }
 
+    /// Dump the network traffic to a file
+    pub fn dump_to_file(&self, path: impl AsRef<Path>) -> Result<(), io::Error> {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-client/src/dvr/record.rs:74]: Nice!

**Review by jdisanti — APPROVED** [review]

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs [hunk 8]
```diff
@@ -11,12 +11,9 @@ pub trait RuntimePlugin {
     fn configure(&self, cfg: &mut ConfigBag) -> Result<(), BoxError>;
 }
 
-impl<T> From<T> for Box<dyn RuntimePlugin>
-where
-    T: RuntimePlugin + 'static,
-{
-    fn from(t: T) -> Self {
-        Box::new(t) as _
+impl RuntimePlugin for Box<dyn RuntimePlugin> {
```
Comment (ysaito1001) [inline rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs:14]: Thanks for updating this.

**Review by ysaito1001 — APPROVED** [review]

**Pull request merged**